keps/sig-api-machinery/storage-hash.md

```diff
@@ -0,0 +1,175 @@
+
+```golang
+type APIResource struct {
+        // A unique ID of the set of objects this resource endpoint represents.
+        // Resources have the same internalID if and only if they refer to the
+        // same set of objects.
+        InternalID string
+	// These are the existing fields.
+	Name string
+	SingularName string
+	Namespaced bool
+	Group string
+	Version string
+	Kind string
+	Verbs Verbs
+	ShortNames []string
+	Categories []string
+}
+```
+
+### Implementation details
+
+For built-in resources, their `internalID`s are set to `SHA256(<etcd key
+prefix>)`. For example, for both `extensions/v1beta1/replicasets` and
+`apps/v1/replicasets`, the etcd key prefix is `/registry/replicasets`. Hashing
+the prefix is to encourage the clients to only test the equality of the hashed
+values, instead of relying on the absolute value.
+
+For custom resources, their `internalID`s are also set to `SHA256(<etcd key
+prefix>)`. In the current implementation, the etcd key prefix for a custom
+resource is `/registry/<crd.spec.group>/<crd.spec.names.plural>`.
+
+For aggregated resources, because their discovery doc is fully controlled
+by the aggregated apiserver, the kube-apiserver has no means to validate their
+`internalID`. If the server is implemented with the generic apiserver library,
+the `internalID` will be `SHA256(<etcd key prefix>)`.
+
+For subresources, the `internalID` field is left empty. If the main resource can
+be accessed via another resource endpoint, then so does the subresource.
+
+For non-persistent resources like `tokenReviews` or `subjectAccessReviews`,
+though the objects are not persisted, the [forward compatibility][] motivation
+still applies, e.g., admins might configure the admission webhooks to intercept
+requests sent to all endpoints. Thus, the `internalID` cannot be left empty, it
+will be set to `SHA256(<the would-be etcd key prefix>)`, e.g., for
+`tokenReviews`, it's `SHA256(/registry/tokenreviews)`. 
+
+[forward compatibility]:#breaking-forwards-compatibility
+
+### Risks and Mitigations
+
+In the future, the "etcd key prefix" might not be sufficient to uniquely
+identify a set of objects. We can always add more factors to the `internalID` to
+ensure their uniqueness. It does not break backwards compatibility because the
+`internalID` is opaque.
+
+Another risk is that an aggregated apiserver accidentally reports `internalID`
+that's identical to the built-in or the custom resources, this will confuse
+clients. Because the kube-apiserver has zero control over the discovery doc of
+aggregated resources, it cannot do any validation to prevent this kind of error.
+It will be aggregated apiserver provider's responsibility to prevent such errors.
+
+## Graduation Criteria
+
+Because the discovery API is a GA feature, to add the `internalID` field, it
+needs to be of GA quality.
+
+## Alternatives
```

Comment:
For built in resources I think we could consider picking a UUID per resource instead of a hash. But I'm not sure how to make that approach work for CRDs.

Reply:
Is the UUID randomly generated at server startup?